dali/operators/reader/webdataset_reader_op.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/webdataset_reader_op.h"
+#include <algorithm>
+#include <cstring>
+#include <string>
+
+namespace dali {
+
+void WebdatasetReader::RunImpl(HostWorkspace& ws) {
+  int num_outputs = ws.NumOutput();
+  int num_samples = GetCurrBatchSize();
+
+  for (int data_idx = 0; data_idx < num_samples; data_idx++) {
+    auto& sample = GetSample(data_idx);
+    for (int output_idx = 0; output_idx < num_outputs; output_idx++) {
+      ws.OutputRef<CPUBackend>(output_idx)[data_idx].Resize(sample[output_idx].shape());
+      ws.OutputRef<CPUBackend>(output_idx)[data_idx].set_type(sample[output_idx].type());
+      ws.OutputRef<CPUBackend>(output_idx)[data_idx].SetMeta(sample[output_idx].GetMeta());
+      std::memcpy(ws.OutputRef<CPUBackend>(output_idx)[data_idx].raw_mutable_data(),
+                  sample[output_idx].raw_data(), sample[output_idx].nbytes());
+    }
+  }
+}
+
+DALI_SCHEMA(readers__Webdataset)
+    .DocStr(
+        R"code(A reader for the webdataset format.
+
+The webdataset format is a dataset meant to improve caching of the data between the samples read.
+The data itself is contained within one or several tar archives, each of which is further split into
+samples, each of which contains one or several components that correspond to the actual file
+contained within the archive. The components that correspond to specific sample are aggregated by
+the part of the filepath that does not correspond to the extension (for the specifics about the
+extensions please read the description of the ``ext`` parameter below). Note that samples with
+their filename starting with a dot will not be loaded, as well as entries that are not a file.
```

Comment:
```suggestion
The webdataset format is a way of providing efficient access to datasets stored in POSIX tar archives.

Storing data in POSIX tar archives greatly speeds up I/O operations on rotational storage and on networked file systems because it permits all I/O operations to operate as large sequential reads and writes.

WebDataset fulfills a similar function to Tensorflow's TFRecord/tf.Example classes, but it is much easier to adopt because it does not actually require any kind of data conversion: data is stored in exactly the same format inside tar files as it is on disk, and all preprocessing and data augmentation code remains unchanged.

The data itself is contained within one or more tar archives, each of which is further split into
samples, each of which contains one or more components that correspond to the actual file
contained within the archive. The components that correspond to specific sample are aggregated by
the part of the filepath that does not correspond to the extension (for the specifics about the
extensions please read the description of the ``ext`` parameter below). Note that samples with
their filename starting with a dot will not be loaded, as well as entries that are not a file.
```

I would also add a reference to the WebDataset documentation.

Reply:
```suggestion
The webdataset format is a way of providing efficient access to datasets stored in POSIX tar archives.

Storing data in POSIX tar archives greatly speeds up I/O operations on mechanical storage devices and on network file systems because it allows the operating system to reduce the number of I/O operations and to read the data ahead.

WebDataset fulfils a similar function to Tensorflow's TFRecord/tf.Example classes, but is much easier to adopt because it does not actually require any data conversion. The data is stored in exactly the same format inside tar files as it is on disk, and all preprocessing and data augmentation code remains unchanged.

The dataset consists of one or more tar archives, each of which is further split into
samples. A sample contains one or more components that correspond to the actual files
contained within the archive. The components that belong to a specific sample are aggregated by
filename without extension (for the specifics about the
extensions please read the description of the ``ext`` parameter below). Note that samples with
their filename starting with a dot will not be loaded, as well as entries that are not regular files.
```